src/github.com/matrix-org/dendrite/clientapi/routing/login.go

```diff
@@ -103,41 +202,10 @@ func Login(
 			}
 		}
 
-		acc, err := accountDB.GetAccountByPassword(req.Context(), localpart, r.Password)
-		if err != nil {
-			// Technically we could tell them if the user does not exist by checking if err == sql.ErrNoRows
-			// but that would leak the existence of the user.
-			return util.JSONResponse{
-				Code: 403,
-				JSON: jsonerror.Forbidden("username or password was incorrect, or the account does not exist"),
-			}
-		}
-
-		token, err := auth.GenerateAccessToken()
-		if err != nil {
-			httputil.LogThenError(req, err)
-		}
-
-		// TODO: Use the device ID in the request
-		dev, err := deviceDB.CreateDevice(
-			req.Context(), acc.Localpart, nil, token, r.InitialDisplayName,
-		)
-		if err != nil {
-			return util.JSONResponse{
-				Code: 500,
-				JSON: jsonerror.Unknown("failed to create device: " + err.Error()),
-			}
-		}
-
-		return util.JSONResponse{
-			Code: 200,
-			JSON: loginResponse{
-				UserID:      dev.UserID,
-				AccessToken: dev.AccessToken,
-				HomeServer:  cfg.Matrix.ServerName,
-				DeviceID:    dev.ID,
-			},
+		if r.Type == PasswordBased {
+			return *handleTokenLogin(r, deviceDB, req, cfg)
```

Comment:
In the project we need to standardize over whether we're returning a `*util.JSONResponse` or `util.JSONResponse`, because in `handleTokenLogin` we're returning a reference, then removing the reference directly afterwards.

It's a bigger issue than just this PR, but for now I would suggest just returning a `util.JSONResponse` in your `handlePasswordLogin` and `handleTokenLogin` functions and remove the `*` on this line and the next.